lib/elixir/lib/map.ex

```diff
@@ -93,11 +95,18 @@ defmodule Map do
   """
   @spec new(Enum.t, (term -> {key, value})) :: map
   def new(enumerable, transform) do
-    fun = fn el, acc ->
-      {k, v} = transform.(el)
-      put(acc, k, v)
-    end
-    Enum.reduce(enumerable, %{}, fun)
+    enumerable
+    |> Enum.to_list
```

Comment:
So my current implementation for a range `0..100_000`. Note this is for mapset at the moment

```
MapSet.new range manual recursion                  100   27522.42 µs/op
MapSet.new range master                             50   54505.60 µs/op
```

I made a version that avoided a second reversal for non list enumerables, but it only got us 3% which is close enough to the margin of error to not bother with the extra code.

Here's the comparison when handed pure lists.

```
MapSet.new list manual recursion                   100   19556.39 µs/op
MapSet.new list master                              50   51277.38 µs/op
```